File: src/services/modules/omp_config.ts
import { request } from "../request";
import type { PiOmpApplyItem, PiOmpProviderMeta } from "@/lib/piOmpCommon";

export type OmpAppType = "omp";

export interface OmpConfigPaths {
  appType: OmpAppType;
  agentDir: string;
  modelsPath: string;
  settingsPath: string;
  profilesDir: string;
  modelsFormat: "json" | "yaml";
  settingsFormat: "json" | "yaml";
  modelsExists: boolean;
  settingsExists: boolean;
}

export interface OmpDefaultSelection {
  provider?: string | null;
  model?: string | null;
  thinkingLevel?: string | null;
  selector?: string | null;
}

export interface OmpWorkspaceState {
  paths: OmpConfigPaths;
  providers: PiOmpProviderMeta[];
  defaultSelection: OmpDefaultSelection;
  modelsText: string;
  settingsText: string;
}

export interface OmpProviderData {
  meta: PiOmpProviderMeta;
  providerJson: unknown;
  providerText: string;
  modelsText: string;
  settingsText: string;
  paths: OmpConfigPaths;
  defaultSelection: OmpDefaultSelection;
}

export interface SaveOmpProviderRequest {
  id: string;
  name?: string;
  enabled: boolean;
  order: number;
  providerJson: unknown;
}

export interface ApplyOmpConfigRequest {
  items: PiOmpApplyItem[];
  defaultProvider?: string | null;
  defaultModel?: string | null;
  thinkingLevel?: string | null;
}

export interface ApplyOmpConfigResult {
  paths: OmpConfigPaths;
  providers: PiOmpProviderMeta[];
  defaultSelection: OmpDefaultSelection;
  enabledCount: number;
}

export const OMP_QUERY_KEYS = {
  workspace: () => ["omp-config", "workspace"] as const,
  provider: (id: string | null) => ["omp-config", "provider", id] as const,
};

export const ompConfigApi = {
  resolvePaths: () => request<OmpConfigPaths>("resolve_omp_config_paths", {}),
  sync: () => request<OmpWorkspaceState>("sync_omp_providers", {}),
  getProvider: (id: string) => request<OmpProviderData>("get_omp_provider", { id }),
  saveProvider: (req: SaveOmpProviderRequest) =>
    request<PiOmpProviderMeta>("save_omp_provider", { req }),
  deleteProvider: (id: string) => request<OmpWorkspaceState>("delete_omp_provider", { id }),
  renameProvider: (oldId: string, newId: string) =>
    request<OmpWorkspaceState>("rename_omp_provider", { oldId, newId }),
  apply: (req: ApplyOmpConfigRequest) => request<ApplyOmpConfigResult>("apply_omp_config", { req }),
};

File: src-tauri/src/models/omp_config.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// omp 配置路径与拆分目录。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OmpConfigPaths {
    pub app_type: String,
    pub agent_dir: String,
    pub models_path: String,
    pub settings_path: String,
    pub profiles_dir: String,
    pub models_format: String,
    pub settings_format: String,
    pub models_exists: bool,
    pub settings_exists: bool,
}

/// omp 当前默认模型选择：`modelRoles.default` = `provider/model[:thinkingLevel]`，含思考档位。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OmpDefaultSelection {
    pub provider: Option<String>,
    pub model: Option<String>,
    pub thinking_level: Option<String>,
    pub selector: Option<String>,
}

/// omp 拆分渠道列表项。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OmpProviderMeta {
    pub id: String,
    pub name: String,
    pub enabled: bool,
    pub order: usize,
    pub model_count: usize,
    pub is_default: bool,
    pub updated_at: String,
    pub configured_at: String,
    pub applied_at: Option<String>,
}

/// omp 页面进入时需要的完整状态。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OmpWorkspaceState {
    pub paths: OmpConfigPaths,
    pub providers: Vec<OmpProviderMeta>,
    pub default_selection: OmpDefaultSelection,
    pub models_text: String,
    pub settings_text: String,
}

/// omp 单个拆分渠道详情。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OmpProviderData {
    pub meta: OmpProviderMeta,
    pub provider_json: Value,
    pub provider_text: String,
    pub models_text: String,
    pub settings_text: String,
    pub paths: OmpConfigPaths,
    pub default_selection: OmpDefaultSelection,
}

/// omp 保存单个拆分渠道请求。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveOmpProviderRequest {
    pub id: String,
    #[serde(default)]
    pub name: Option<String>,
    pub enabled: bool,
    pub order: usize,
    pub provider_json: Value,
}

/// omp 应用请求（含 thinking_level）。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyOmpConfigRequest {
    pub items: Vec<crate::modules::tool_config::pi_omp_common::ApplyItem>,
    #[serde(default)]
    pub default_provider: Option<String>,
    #[serde(default)]
    pub default_model: Option<String>,
    #[serde(default)]
    pub thinking_level: Option<String>,
}

/// omp 应用结果。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyOmpConfigResult {
    pub paths: OmpConfigPaths,
    pub providers: Vec<OmpProviderMeta>,
    pub default_selection: OmpDefaultSelection,
    pub enabled_count: usize,
}
